### ISSUE TITLE
Realtime: Presence

### ISSUE THREAD

**Issue by yavor87** [ably/ably-dotnet#8]
- [ ]   **RTP1** When a channel `ATTACHED` `ProtocolMessage` is received, the `ProtocolMessage` may contain a bit flag with value 1 indicating that there are currently members present on the channel. If members are present, subsequent `ProtocolMessage SYNC` messages will be delivered with presence members (messages) until the `SYNC` operation is complete. If there is no flag or the right most bit is zero, then the presence map should be considered in sync immediately as there are no members present on the channel at the time of attach
- [x]   **RTP2** A [PresenceMap](https://github.com/ably/ably-java/blob/master/src/io/ably/realtime/Presence.java#L384-L529) should be used to maintain a list of members present on a channel. As there are no guarantees that during the `SYNC` phase presence events will arrive in order i.e. a leave event for a member can arrive before that member is later registered as as present as part of the initial `SYNC` operation. As such, until the `SYNC` operation is complete, timestamps must be recorded for all members to ensure the most recent present state is used, see the [Java implementation](https://github.com/ably/ably-java/blob/master/src/io/ably/realtime/Presence.java#L418-L431). Once a `SYNC` event is complete, the members that are now considered `ABSENT` can be removed from the map, see the [Java implementation](https://github.com/ably/ably-java/blob/master/src/io/ably/realtime/Presence.java#L504-L509).
- [ ]   **RTP3** If a `SYNC` operation is underway but not yet complete, and the connection is disconnected unexpectedly, then if the connection is resumed successfully, it is the responsibility of the client library to complete the `SYNC` operation. The client library requests a `SYNC` resume by sending a `SYNC` `ProtocolMessage` with the last received sync serial number. See the [Ruby implementation](https://github.com/ably/ably-ruby/blob/7b18a20/lib/ably/realtime/presence/members_map.rb#L169-L176) and the [Ruby test](https://github.com/ably/ably-ruby/blob/7b18a20/spec/acceptance/realtime/presence_spec.rb#L1338-L1357)
- [ ]   **RTP4** Ensure a test exists that enters 250 members using `Presence#enterClient` on a single connection, and checks for `PRESENT` events to be emitted on another connection for each member, and once sync is complete, all 250 members should be present in a `Presence#get` request
- [x]   **RTP5** Channel state change side effects:
  - [x]   **RTP5a** If the channel enters the `DETACHED` or `FAILED` state then all queued presence messages will fail immediately
  - [x]   **RTP5b** If a channel enters the `ATTACHED` state then all queued presence messages will be sent immediately and a presence `SYNC` will be initiated implicitly
- [ ]   **RTP6** `Presence#subscribe` function:
  - [x]   **RTP6a** Subscribe with no arguments subscribes a listener to all presence messages
  - [ ]   **RTP6b** Subscribe with a single action argument – such as `ENTER`, `LEAVE`, `UPDATE` or `PRESENT` – subscribes a listener to receive only presence messages with that action
  - [ ]   **RTP6c** Implicitly attaches the channel if not attached
- [ ]   **RTP7** `Presence#unsubscribe` function:
  - [x]   **RTP7a** Unsubscribe with no arguments unsubscribes the listener if previously subscribed with an action-specific subscription
  - [ ]   **RTP7b** Unsubscribe with a single action argument unsubscribes the provided listener to all presence messages for that action
- [ ]   **RTP8** `Presence#enter` function:
  - [x]   **RTP8a** Enters this client into this channel
  - [x]   **RTP8b** Optionally a callback can be provided that is called for both success or failure to enter
  - [x]   **RTP8c** A `PRESENCE ProtocolMessage` with a `PresenceMessage` with the action `ENTER` is sent to the Ably service
  - [x]   **RTP8d** Implicitly attaches to the channel if not attached
  - [ ]   **RTP8e** Optional data can be included when entering a channel that will be encoded / decoded as with normal messages. A test should exist to ensure data used with enter is encoded & decoded correctly. Also, when data is provided when entering, but no data is provided when leaving, the data attribute should be emitted in the `LEAVE` event for this client
  - [ ]   **RTP8f** Raises an exception of the client library has not `clientId` configured
  - [x]   **RTP8g** Raises an exception if the channel is `DETACHED` or `FAILED`
  - [ ]   **RTP8h** Fails and emits an error if the client does not have required presence permissions
- [ ]   **RTP9** `Presence#update` function:
  - [ ]   **RTP9a** Updates the data for the present member with a value or `Null`
  - [ ]   **RTP9b** If the client was not already entered, it enters this client into this channel
  - [x]   **RTP9c** Optionally a callback can be provided that is called for both success or failure to update
  - [x]   **RTP9d** A `PRESENCE ProtocolMessage` with a `PresenceMessage` with the action `UPDATE` is sent to the Ably service
  - [x]   **RTP9e** In all other ways, this method is identical to `Presence#enter` and should have matching tests
- [ ]   **RTP10a** `Presence#leave` function:
  - [ ]   **RTP10b** Leaves this client from the channel and the data will be updated with the value provided. If the language permits the data argument to be omitted, then the previously set data value will be omitted as a convenience
  - [x]   **RTP10c** Optionally a callback can be provided that is called for both success or failure to leave
  - [x]   **RTP10d** A `PRESENCE ProtocolMessage` with a `PresenceMessage` with the action `LEAVE` is sent to the Ably service
  - [ ]   **RTP10e** If the client is not currently `ENTERED` then an exception is raised
  - [x]   **RTP10f** In all other ways, this method is identical to `Presence#enter` and should have matching tests
- [ ]   **RTP11** `Presence#get` function:
  - [x]   **RTP11a** Returns the list of current members on the channel in a callback and, by default, will not wait for the `SYNC` to be completed
  - [ ]   **RTP11b** Raises an exception if the channel is `DETACHED` or `FAILED`
  - [ ]   **RTP11c** An optional set of params can be provided:
    - [ ]   **RTP11c1** `waitForSync` when true, will wait until `SYNC` is complete before returning a list of members
    - [ ]   **RTP11c2** `clientId` filters members by the provided `clientId`
    - [ ]   **RTP11c3** `connectionId` filters members by the provided `connectionId`
- [ ]   **RTP12** `Presence#history` function:
  - [ ]   **RTP12a** Supports all the same params as REST `Presence#history`
  - [ ]   **RTP12b** Additionally supports the param `untilAttach`, which if true, will only retrive messages up to the moment that the channel was attached. This bound is specified by passing the querystring param `from_serial` with the serial number assigned to the channel in the `ATTACHED` `ProtocolMessage`. If the `untilAttach` param is specified when the channel is not attached an exception should be raised.
  - [ ]   **RTP12c** Returns a `PaginatedResult` page containing the first page of messages in the `PaginatedResult#items` attribute returned from the history request.
  - [ ]   **RTP12d** A test should exist that registers presence with a few clients, and upon confirmation of entering the channel for all clients, a presence history request should be made using another client to ensure all presence events are available
- [ ]   **RTP13** `Presence#syncComplete` returns true if the initial `SYNC` operation has completed for the members present on the channel
- [ ]   **RTP14** `Presence#enterClient` function:
  - [x]   **RTP14a** Enters into presence on a channel on behalf of another `clientId`. This allows a single client with suitable permissions to register presence on behalf of any number of clients using a single connection
  - [x]   **RTP14b** Optionally a callback can be provided that is called for both success or failure to enter
  - [ ]   **RTP14c** Data can optionally be provided when entering and will follow the normal encoding & decoding rules
  - [ ]   **RTP14d** A test should exist that registers a number of members each with different @clientId@s on a presence channel, and then a `Presence#get` should be used to verify that all members are present as expected
- [ ]   **RTP15** `Presence#enterClient` `Presence#updateClient` and `Presence#leaveClient` function:
  - [x]   **RTP15a** Performs an enter, update or leave for given `clientId`. These methods apply if the Realtime library was not initialised with a specific `clientId`. This allows a single client with suitable permissions to update presence on behalf of any number of clients using a single connection. Otherwise these are functionality equivalent to the corresponding `enter`, `update` and `leave` methods, and equivalent test coverage should be provided
  - [ ]   **RTP15b** Tests should use `enterClient`, `updateClient` and `leaveClient` for many members from one `Realtime` client and check that the operations are reflected in the presence map and the expected events are emitted on a separate client.
  - [ ]   **RTP15c** Tests should also ensure that using these methods has no side effects on a client that has entered normally using `Presence#enter`
  - [x]   **RTP15d** A callback can be provided that will be called upon success or failure
  - [x]   **RTP15e** A channel will be implicitly attached when these methods are called

**Comment by mattheworiordan** [comment]
Prioritisation notes:

### High priority
* `SYNC` support assuming the connection did not disconnect during `SYNC`
* `Presence#enter`, `Presence#update`, `Presence#leave` with all options (no encryption)
* `Presence#subscribe` and `Presence#unsubscribe`
* `Presence#get` using `PresenceMap`
* `Presence#history`

### Medium priority
* Full support for `SYNC`, including recovery if underlying transport disconnected and reconnected during a SYNC i.e. the client explicitly requests a new SYNC

### Low priority
* `Presence#enterClient`, `Presence#updateClient`, `Presence#leaveClient`
* Encryption

**Comment by marto83** [comment]
Closing as not relevant any more
